### PR TITLE
refactor: use pagination to load values from pstor

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
### Changes in the PR.

- Changes the spec population from etcd from non paged call to a paged call.
- Also incorporates the pagination to legacy migration flows.

### CODE REVIEW

File: utils/pstor/src/etcd.rs [hunk 1]
```diff
@@ -233,7 +233,6 @@ impl StoreKv for Etcd {
                 Some(
                     GetOptions::new()
                         .with_prefix()
-                        .with_from_key()
```
Comment (Abhinandan-Purkait) [inline utils/pstor/src/etcd.rs:236]: Had to remove this otherwise it returns all keys.

Comment (tiagolobocastro) [reply utils/pstor/src/etcd.rs:236]: hmm right but then also without it it won't work right? it's not paginating?

Comment (tiagolobocastro) [reply utils/pstor/src/etcd.rs:236]: Yeah I don't think this is right either.
I think what we need to do is add the range end, so probably we need to pass the original prefix on every call, or compute the range end ahead of time.
Please add a test with small limits to ensure we get it right

Comment (Abhinandan-Purkait) [reply utils/pstor/src/etcd.rs:236]: For a specific prefix, for example, `/a/b/c/NodeSpec/id` all the prefix for these kvs will always be `/a/b/c/NodeSpec` how do we calculate range end?

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/resources/migration.rs [hunk 2]
```diff
@@ -16,31 +19,58 @@ pub const MSP_OPERATOR: &str = "msp-operator";
...
+                    &new_value,
+                )
+                .await?;
+        }
+
+        if let Some((key, _)) = store_entries.last() {
+            prefix = key;
+        } else {
+            break;
+        }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/resources/migration.rs]:
```suggestion
        last = store_entries.last().cloned();
        for (k, v) in store_entries {
            let id = k
                .split('/')
                .last()
                .ok_or_else(|| StoreError::InvalidKey { key: k.to_string() })?;

            let new_value = match spec_type {
                StorableObjectType::VolumeSpec => migrate_volume_labels(v)?,
                StorableObjectType::PoolSpec => migrate_pool_labels(v)?,
                _ => v,
            };

            store
                .put_kv(
                    &format!("{}/{}", key_prefix_obj(spec_type, API_VERSION), id),
                    &new_value,
                )
                .await?;
        }

        if let Some((key, _)) = last {
            prefix = key;
        } else {
            break;
        }
```

File: utils/pstor/src/etcd.rs [hunk 3]
```diff
@@ -256,6 +261,42 @@ impl StoreKv for Etcd {
...
+
+            first = false;
+
+            all_values.extend(values.clone());
+
+            if let Some((key, _)) = values.last() {
+                prefix = key;
+            } else {
+                break;
+            }
```
Comment (tiagolobocastro) [inline utils/pstor/src/etcd.rs]:
```suggestion
            last = values.last().cloned();
            all_values.extend(values);

            if let Some((key, _)) = last() {
                prefix = key;
            } else {
                break;
            }
```

File: control-plane/agents/src/bin/core/tests/controller/mod.rs [hunk 4]
```diff
@@ -212,3 +215,76 @@ async fn core_agent_lease_lock() {
...
+    }
+
+    // Persist some volumes in old key space in etcd.
+    for _i in 1 .. 6 {
+        let uuid = Uuid::new_v4();
+        let key = format!("{}/{}", OLD_VOLUME_PREFIX, uuid);
+        let json_str = r#"{"uuid":"456122b1-7e19-4148-a890-579ca785a119","size":2147483648,"labels":{"local":"true"},"num_replicas":3,"status":{"Created":"Online"},"target":{"node":"mayastor-node4","nexus":"d6ccbb97-d13e-4ffb-91a0-c7607bb01f8f","protocol":"nvmf"},"policy":{"self_heal":true},"topology":{"node":{"Explicit":{"allowed_nodes":["mayastor-node2","mayastor-master","mayastor-node3","mayastor-node1","mayastor-node4"],"preferred_nodes":["mayastor-node2","mayastor-node3","mayastor-node4","mayastor-master","mayastor-node1"]}},"pool":{"Labelled":{"exclusion":{},"inclusion":{"openebs.io/created-by":"msp-operator"}}}},"last_nexus_id":"d6ccbb97-d13e-4ffb-91a0-c7607bb01f8f","operation":null}"#;
+        let value = Value::from_str(json_str).unwrap();
+        etcd.put_kv(&key, &value).await.unwrap();
+    }
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/tests/controller/mod.rs:266]: maybe also add the old nexus style entries? $uuid

File: utils/pstor/src/etcd.rs [hunk 5]
```diff
@@ -221,23 +221,28 @@ impl StoreKv for Etcd {
...
         key_prefix: &str,
         limit: i64,
+        range_end: &str,
     ) -> Result<Vec<(String, Value)>, Error> {
         if limit <= 2 {
             return Err(Error::PagedMinimum);
         }
 
+        let get_options = if range_end.is_empty() {
+            println!("range_end is empty");
```
Comment (tiagolobocastro) [inline utils/pstor/src/etcd.rs]:
```suggestion
```

File: utils/pstor/src/etcd.rs [hunk 6]
```diff
@@ -221,23 +221,28 @@ impl StoreKv for Etcd {
...
+            GetOptions::new()
+                .with_from_key()
```
Comment (tiagolobocastro) [inline utils/pstor/src/etcd.rs:232]: what does this do in this case?

**Review by tiagolobocastro — APPROVED** [review]

File: utils/pstor/src/etcd.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply utils/pstor/src/etcd.rs]: Key becomes string with this, causes lifetime issue?

File: utils/pstor/src/etcd.rs (hunk 6 above)
Comment (Abhinandan-Purkait) [reply utils/pstor/src/etcd.rs:232]: If the range end is empty then it `Gets all keys >= key.`

Comment (Abhinandan-Purkait) [reply utils/pstor/src/etcd.rs:232]: So for example if its `get "" ""` it will get us everything paginated

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**